File: frontend/src/types/channel.ts
// src/types/channel.ts
// Types matching backend Pydantic models (app/models/channel.py)
// Design doc: CHANNEL_MODEL_DESIGN.md v1.1.0

// === Enums ===

export type BrokerType =
  // Tier 1 — Core
  | "kafka"
  | "redpanda"
  | "rabbitmq"
  | "pulsar"
  | "nats"
  | "google_pubsub"
  | "aws_sns_sqs"
  | "azure_servicebus"
  | "redis_streams"
  // Tier 2 — Enterprise & IoT
  | "solace"
  | "ibmmq"
  | "activemq_artemis"
  | "mqtt"
  | "mqtt_secure"
  | "websocket"
  | "websocket_secure"
  | "anypoint_mq"
  | "mercure"
  | "stomp"
  // Tier 3 — Serverless
  | "amazon_kinesis"
  | "amazon_eventbridge"
  | "custom";

export type ResourceKind = "topic" | "exchange" | "subject" | "queue" | "stream" | "channel" | "destination" | "event_bus";

export type MessagingPattern = "topic_log" | "pubsub" | "queue" | "request_reply" | "broadcast" | "event_bus";

export type BindingStrategy = "channel_bound" | "domain_bound" | "app_bound";

export type BindingOrigin =
  | "tns"
  | "trs"
  | "rns_heuristic"
  | "kafka_api"
  | "routing_key"
  | "attribute_filter"
  | "manual";

export type BindingStatus = "active" | "missing_subject" | "stale" | "unverified";

export type SchemaRole = "value" | "key" | "header" | "envelope";

// Re-export from governance (canonical location)
export type { DataLayer } from "@/types/governance";

// === Channels ===

export interface ChannelCreate {
  name: string;
  address: string;
  broker_type: BrokerType;
  resource_kind: ResourceKind;
  messaging_pattern: MessagingPattern;
  broker_config?: Record<string, unknown>;
  data_layer?: string | null;
  description?: string | null;
  owner?: string | null;
  tags?: string[];
}

export interface ChannelUpdate {
  name?: string | null;
  address?: string | null;
  broker_type?: BrokerType | null;
  resource_kind?: ResourceKind | null;
  messaging_pattern?: MessagingPattern | null;
  broker_config?: Record<string, unknown> | null;
  data_layer?: string | null;
  description?: string | null;
  owner?: string | null;
  tags?: string[] | null;
}

export interface ChannelResponse {
  id: string;
  registry_id: string;
  name: string;
  address: string;
  broker_type: BrokerType;
  resource_kind: ResourceKind;
  messaging_pattern: MessagingPattern;
  broker_config: Record<string, unknown>;
  data_layer: string | null;
  description: string | null;
  owner: string | null;
  tags: string[];
  is_auto_detected: boolean;
  auto_detect_source: string | null;
  created_at: string | null;
  updated_at: string | null;
  subjects: ChannelSubjectResponse[];
}

export interface ChannelSummary {
  id: string;
  name: string;
  address: string;
  broker_type: BrokerType;
  resource_kind: ResourceKind;
  messaging_pattern: MessagingPattern;
  data_layer: string | null;
  subject_count: number;
  has_key_schema: boolean;
  has_value_schema: boolean;
  binding_health: string;
}

// === Bindings ===

export interface ChannelSubjectCreate {
  subject_name: string;
  binding_strategy: BindingStrategy;
  schema_role?: SchemaRole;
  binding_origin?: BindingOrigin;
  binding_selector?: string | null;
}

export interface ChannelSubjectResponse {
  id: string;
  channel_id: string;
  subject_name: string;
  binding_strategy: BindingStrategy;
  schema_role: SchemaRole;
  binding_origin: BindingOrigin;
  binding_selector: string | null;
  binding_status: BindingStatus;
  last_verified_at: string | null;
  is_auto_detected: boolean;
  created_at: string | null;
}

// === Aggregated views ===

export interface ChannelMapEntry {
  channel: ChannelSummary;
  bindings: ChannelSubjectResponse[];
}

export interface ChannelMapResponse {
  channels: ChannelMapEntry[];
  total_channels: number;
  total_bindings: number;
  bound_subjects: number;
  unbound_subjects: string[];
  warnings: string[];
}

// === Display helpers ===

export const BROKER_ICONS: Record<string, string> = {
  // Tier 1
  kafka: "🔶", redpanda: "🐼", rabbitmq: "🐰", pulsar: "⚡", nats: "🔷",
  google_pubsub: "☁️", aws_sns_sqs: "📦", azure_servicebus: "🔵",
  redis_streams: "🔴",
  // Tier 2
  solace: "🌊", ibmmq: "🏢", activemq_artemis: "🏹",
  mqtt: "📡", mqtt_secure: "🔐",
  websocket: "🌐", websocket_secure: "🔒",
  anypoint_mq: "🔀", mercure: "📢", stomp: "🦶",
  // Tier 3
  amazon_kinesis: "🌊", amazon_eventbridge: "🚌", custom: "⚙️",
};

export const BROKER_LABELS: Record<string, string> = {
  // Tier 1
  kafka: "Kafka", redpanda: "Redpanda", rabbitmq: "RabbitMQ",
  pulsar: "Pulsar", nats: "NATS",
  google_pubsub: "Google Pub/Sub", aws_sns_sqs: "AWS SNS/SQS",
  azure_servicebus: "Azure Service Bus", redis_streams: "Redis Streams",
  // Tier 2
  solace: "Solace PubSub+", ibmmq: "IBM MQ", activemq_artemis: "ActiveMQ Artemis",
  mqtt: "MQTT", mqtt_secure: "MQTT (TLS)",
  websocket: "WebSocket", websocket_secure: "WebSocket (TLS)",
  anypoint_mq: "Anypoint MQ", mercure: "Mercure (SSE)", stomp: "STOMP",
  // Tier 3
  amazon_kinesis: "Amazon Kinesis", amazon_eventbridge: "Amazon EventBridge",
  custom: "Custom",
};


export const PATTERN_LABELS: Record<string, string> = {
  topic_log: "Topic Log",
  pubsub: "Pub/Sub",
  queue: "Queue",
  request_reply: "Request/Reply",
  broadcast: "Broadcast",
  event_bus: "Event Bus",
};

export const RESOURCE_LABELS: Record<ResourceKind, string> = {
  topic: "Topic", exchange: "Exchange", subject: "Subject",
  queue: "Queue", stream: "Stream", channel: "Channel",
  destination: "Destination", event_bus: "Event Bus",
};

export const STRATEGY_LABELS: Record<BindingStrategy, string> = {
  channel_bound: "Channel-Bound",
  domain_bound: "Domain-Bound",
  app_bound: "App-Bound",
};

export const STATUS_CONFIG: Record<BindingStatus, { label: string; color: string; dot: string }> = {
  active: { label: "Active", color: "text-emerald-400", dot: "bg-emerald-400" },
  missing_subject: { label: "Missing", color: "text-red-400", dot: "bg-red-400" },
  stale: { label: "Stale", color: "text-amber-400", dot: "bg-amber-400" },
  unverified: { label: "Unverified", color: "text-slate-400", dot: "bg-slate-400" },
};

export const DEFAULT_RESOURCE: Record<BrokerType, ResourceKind> = {
  // Tier 1
  kafka: "topic", redpanda: "topic", rabbitmq: "exchange",
  pulsar: "topic", nats: "subject",
  google_pubsub: "topic", aws_sns_sqs: "queue",
  azure_servicebus: "queue", redis_streams: "stream",
  // Tier 2
  solace: "topic", ibmmq: "queue", activemq_artemis: "queue",
  mqtt: "topic", mqtt_secure: "topic",
  websocket: "channel", websocket_secure: "channel",
  anypoint_mq: "queue", mercure: "topic", stomp: "destination",
  // Tier 3
  amazon_kinesis: "stream", amazon_eventbridge: "event_bus", custom: "topic",
};

export const DEFAULT_PATTERN: Record<string, MessagingPattern> = {
    // Tier 1
    kafka: "topic_log", redpanda: "topic_log", rabbitmq: "pubsub",
    pulsar: "topic_log", nats: "pubsub",
    google_pubsub: "pubsub", aws_sns_sqs: "queue",
    azure_servicebus: "queue", redis_streams: "topic_log",
    // Tier 2
    solace: "pubsub", ibmmq: "queue", activemq_artemis: "queue",
    mqtt: "pubsub", mqtt_secure: "pubsub",
    websocket: "broadcast", websocket_secure: "broadcast",
    anypoint_mq: "queue", mercure: "broadcast", stomp: "queue",
    // Tier 3
    amazon_kinesis: "topic_log", amazon_eventbridge: "event_bus",
    custom: "topic_log",
};

File: backend/app/models/channel.py
"""
event7 - Channel Model
Modèles pour l'abstraction Channel (multi-broker, multi-protocol).

Placement: backend/app/models/channel.py

Design doc: CHANNEL_MODEL_DESIGN.md v1.1.0
- 7 enums (BrokerType, ResourceKind, MessagingPattern, BindingStrategy,
           BindingOrigin, BindingStatus, DataLayer, SchemaRole)
- 6 Pydantic models (ChannelCreate, ChannelResponse, ChannelSubjectCreate,
                      ChannelSubjectResponse, ChannelSummary, ChannelMapResponse)
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


# ====================================================================
# ENUMS
# ====================================================================


class BrokerType(str, Enum):
    """Type de broker / système de messaging."""
    # Tier 1 — Core
    KAFKA = "kafka"
    REDPANDA = "redpanda"
    RABBITMQ = "rabbitmq"
    PULSAR = "pulsar"
    NATS = "nats"
    GOOGLE_PUBSUB = "google_pubsub"
    AWS_SNS_SQS = "aws_sns_sqs"
    AZURE_SERVICEBUS = "azure_servicebus"
    REDIS_STREAMS = "redis_streams"
    # Tier 2 — Enterprise & IoT
    SOLACE = "solace"
    IBMMQ = "ibmmq"
    ACTIVEMQ_ARTEMIS = "activemq_artemis"
    MQTT = "mqtt"
    MQTT_SECURE = "mqtt_secure"
    WEBSOCKET = "websocket"
    WEBSOCKET_SECURE = "websocket_secure"
    ANYPOINT_MQ = "anypoint_mq"
    MERCURE = "mercure"
    STOMP = "stomp"
    # Tier 3 — Serverless
    AMAZON_KINESIS = "amazon_kinesis"
    AMAZON_EVENTBRIDGE = "amazon_eventbridge"
    CUSTOM = "custom"


class ResourceKind(str, Enum):
    """Nature physique de la ressource broker derrière un Channel."""
    TOPIC = "topic"
    EXCHANGE = "exchange"
    SUBJECT = "subject"
    QUEUE = "queue"
    STREAM = "stream"
    CHANNEL = "channel"
    DESTINATION = "destination"
    EVENT_BUS = "event_bus"

class MessagingPattern(str, Enum):
    """Pattern de messaging du channel."""
    TOPIC_LOG = "topic_log"     # Kafka, Pulsar, Redis Streams
    PUBSUB = "pubsub"           # Pub/Sub, SNS, NATS
    QUEUE = "queue"             # RabbitMQ, SQS, Azure SB


class BindingStrategy(str, Enum):
    """Stratégie conceptuelle de liaison subject ↔ channel.
    Décrit la logique métier du lien, pas son mode de détection."""
    CHANNEL_BOUND = "channel_bound"   # ≈ TNS — schema couplé au channel
    DOMAIN_BOUND = "domain_bound"     # ≈ RNS — schema modèle métier
    APP_BOUND = "app_bound"           # ≈ RNS scopé app


class BindingOrigin(str, Enum):
    """Mode concret par lequel le binding a été établi.
    Distinct de la strategy (conceptuelle). Permet la traçabilité."""
    TNS = "tns"
    TRS = "trs"
    RNS_HEURISTIC = "rns_heuristic"
    KAFKA_API = "kafka_api"
    ROUTING_KEY = "routing_key"
    ATTRIBUTE_FILTER = "attribute_filter"
    MANUAL = "manual"


class BindingStatus(str, Enum):
    """État de santé du binding — le subject référencé existe-t-il encore ?"""
    ACTIVE = "active"
    MISSING_SUBJECT = "missing_subject"
    STALE = "stale"
    UNVERIFIED = "unverified"


class DataLayer(str, Enum):
    """Layer de maturité de la donnée.
    Layer primaire = sur le subject/enrichment (sémantique de donnée).
    Layer sur le channel = hint UX dérivé (layer dominant attendu du flux)."""
    RAW = "raw"
    CORE = "core"
    REFINED = "refined"
    APPLICATION = "application"


class SchemaRole(str, Enum):
    """Rôle du schema dans le channel."""
    VALUE = "value"
    KEY = "key"
    HEADER = "header"
    ENVELOPE = "envelope"


# ====================================================================
# PYDANTIC MODELS — Channels
# ====================================================================


class ChannelCreate(BaseModel):
    """Payload pour créer un channel."""
    name: str = Field(..., min_length=1, max_length=500, examples=["Billing Events"])
    address: str = Field(..., min_length=1, max_length=1000, examples=["corp.billing.events.v1"])
    broker_type: BrokerType
    resource_kind: ResourceKind
    messaging_pattern: MessagingPattern
    broker_config: dict = Field(default_factory=dict)
    data_layer: DataLayer | None = None
    description: str | None = None
    owner: str | None = Field(default=None, max_length=200)
    tags: list[str] = Field(default_factory=list)


class ChannelUpdate(BaseModel):
    """Payload pour mettre à jour un channel (partiel)."""
    name: str | None = Field(default=None, max_length=500)
    address: str | None = Field(default=None, max_length=1000)
    broker_type: BrokerType | None = None
    resource_kind: ResourceKind | None = None
    messaging_pattern: MessagingPattern | None = None
    broker_config: dict | None = None
    data_layer: DataLayer | None = None
    description: str | None = None
    owner: str | None = None
    tags: list[str] | None = None


class ChannelResponse(BaseModel):
    """Channel retourné par l'API."""
    id: str
    registry_id: str
    name: str
    address: str
    broker_type: BrokerType
    resource_kind: ResourceKind
    messaging_pattern: MessagingPattern
    broker_config: dict = Field(default_factory=dict)
    data_layer: DataLayer | None = None
    description: str | None = None
    owner: str | None = None
    tags: list[str] = Field(default_factory=list)
    is_auto_detected: bool = False
    auto_detect_source: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    # Relations (peuplé par le service)
    subjects: list["ChannelSubjectResponse"] = Field(default_factory=list)


class ChannelSummary(BaseModel):
    """Vue légère pour les listings."""
    id: str
    name: str
    address: str
    broker_type: BrokerType
    resource_kind: ResourceKind
    messaging_pattern: MessagingPattern
    data_layer: DataLayer | None = None
    subject_count: int = 0
    has_key_schema: bool = False
    has_value_schema: bool = False
    binding_health: str = "unknown"


# ====================================================================
# PYDANTIC MODELS — Channel-Subject Bindings
# ====================================================================


class ChannelSubjectCreate(BaseModel):
    """Payload pour lier un subject à un channel."""
    subject_name: str = Field(..., min_length=1, max_length=500)
    binding_strategy: BindingStrategy
    schema_role: SchemaRole = SchemaRole.VALUE
    binding_origin: BindingOrigin = BindingOrigin.MANUAL
    binding_selector: str | None = Field(
        default=None,
        max_length=500,
        description="Routing key pattern, attribute filter, record name, etc.",
        examples=["billing.invoice.*", "eventType=InvoiceCreated"],
    )


class ChannelSubjectResponse(BaseModel):
    """Binding retourné par l'API."""
    id: str
    channel_id: str
    subject_name: str
    binding_strategy: BindingStrategy
    schema_role: SchemaRole
    binding_origin: BindingOrigin
    binding_selector: str | None = None
    binding_status: BindingStatus = BindingStatus.UNVERIFIED
    last_verified_at: datetime | None = None
    is_auto_detected: bool = False
    created_at: datetime | None = None


# ====================================================================
# PYDANTIC MODELS — Aggregated views
# ====================================================================


class ChannelMapEntry(BaseModel):
    """Une entrée dans la vue channel-map (channel + ses bindings)."""
    channel: ChannelSummary
    bindings: list[ChannelSubjectResponse] = Field(default_factory=list)


class ChannelMapResponse(BaseModel):
    """Vue complète channel-map pour le frontend."""
    channels: list[ChannelMapEntry] = Field(default_factory=list)
    total_channels: int = 0
    total_bindings: int = 0
    bound_subjects: int = 0
    unbound_subjects: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)